composer/datasets/streaming/writer.py

```diff
@@ -82,29 +88,63 @@ def __init__(self, dirname: str, fields: List[str], shard_size_limit: int = 1 <<
         self.new_samples = []
         self.new_shard_size = 0
 
+        # compression scheme for shards
+        self.compression_scheme = None
+        self.open_f = lambda n: open(n, 'xb')
+        self._set_compression_settings(compression)
+
+    def _set_compression_settings(self, compression: Optional[str]):
+        """Sets compression settings for the given compression algorithm
+
+        Args:
+            compression (str): Compression algorithm and optional compression level. Currently supported: 'gz', 'gz:[1-9]' or None.
+        """
+        gz_compression_level = 6
+        if compression is not None:
+            compression_args = compression.split(':')
+            self.compression_scheme = compression_args[0]
+            if self.compression_scheme == 'gz' and len(compression_args) > 1:
+                gz_compression_level = int(compression_args[1])
+
+        else:
+            self.compression_scheme = None
+
+        if self.compression_scheme == 'gz':
+            self.open_f = lambda n: gz.open(n, 'xb', compresslevel=gz_compression_level)
+
     def _flush_shard(self) -> None:
         """Flush cached samples to a new dataset shard."""
         shard = len(self.samples_per_shard)
-        basename = get_shard_basename(shard)
+        basename = get_shard_basename(shard, compression_scheme=self.compression_scheme)
         filename = os.path.join(self.dirname, basename)
-        with open(filename, 'xb') as out:
+
+        with self.open_f(filename) as out:
             for data in self.new_samples:
                 out.write(data)
+
         self.samples_per_shard.append(len(self.new_samples))
         self.bytes_per_shard.append(self.new_shard_size)
         self.new_samples = []
         self.new_shard_size = 0
 
+    def _write_compression_scheme(self) -> None:
```

Comment:
Curious why do we need to store the compression metadata in a separate file?

Reply:
if we store it in the index file, we wouldn't be able to compress the index file.

Reply:
Hmm, could we have the user pass the path to the index file (including the compression extension, if any), rather than the path to the folder?

Reply:
> Anything we can do to avoid adding more args to the StreamingDataset init signature would be much appreciated :)
CC @abhi-mosaic I know we're hesitant to add/change the arguments to the streaming dataset function, what do you think of this?